src/Unit/Algebra.hs

```diff
@@ -34,73 +35,41 @@ import Unit
+u .*. v = pure (prj v * prj u)
 
 infixl 7 .*.
 
-(./.) :: forall dimu u dimv v a . (Mul dimu u dimv (Inv v), Unit dimv v, Fractional a) => u a -> v a -> Res u (Inv v) a
-u ./. v = u .*. Inv @v (1/prj v)
+(./.) :: (Unit du u, Unit dv v, Unit dw (Mul u (Inv v)), Fractional a) => u a -> v a -> Mul u (Inv v) a
+u ./. v = pure (prj u / prj v)
 
 infixl 7 ./.
 
-instance (MulBy step dimu u dimv v, Step u v ~ step, Unit dimu u, Unit dimv v) => Mul dimu u dimv v where
-  type Res u v = ResBy (Step u v) u v
-  (.*.) = mulBy @step
-
-
-class (Unit dimu u, Unit dimv v) => MulBy (step :: Act) dimu u dimv v where
-  type ResBy step u v :: * -> *
-  mulBy :: Fractional a => u a -> v a -> ResBy step u v a
-
--- | Append at the head of the chain.
-instance {-# OVERLAPPABLE #-} (Unit dimu u, Unit dimv v) => MulBy 'Prepend dimu u dimv v where
-  type ResBy 'Prepend u v = u :*: v
-  u `mulBy` v = Prd (prj u * prj v)
-
--- | Elimination by reciprocals at left.
-instance {-# OVERLAPPABLE #-} (Unit dimu u, Unit dimv v, Unit dimu' u', u' ~ InvOf u) => MulBy 'CancelL (dimu :*: dimv) (u :*: v) dimu' u' where
-  type ResBy 'CancelL (u :*: v) u' = v
-  u `mulBy` v = pure (prj u * prj v)
-
--- | Elimination by reciprocals at right.
-instance {-# OVERLAPPABLE #-} (Unit dimu u, Unit dimv v, Unit dimv' v', v' ~ InvOf v) => MulBy 'CancelR (dimu :*: dimv) (u :*: v) dimv' v' where
-  type ResBy 'CancelR (u :*: v) v' = u
-  u `mulBy` v = pure (prj u * prj v)
-
--- | Decompose products on the right.
-instance {-# OVERLAPPABLE #-} (Mul dimu u dimv' v', Mul dimuv' (Res u v') dimv v, Unit dimv' v', Unit dimuv'v (Res (Res u v') v)) => MulBy 'Decompose dimu u (dimv :*: dimv') (v :*: v') where
-  type ResBy 'Decompose u (v :*: v') = Res (Res u v') v
-  u `mulBy` Prd v = pure (prj u * v)
-
--- | Walk the chain.
-instance {-# OVERLAPPABLE #-} (Mul dimu u dimv v, Unit dimu' u') => MulBy 'Walk (dimu :*: dimu') (u :*: u') dimv v where
-  type ResBy 'Walk (u :*: u') v = Res u v :*: u'
-  Prd u `mulBy` v = Prd (u * prj v)
-
+type family Mul u v where
+  Mul  u                  I         = u                           -- u * 1       = u
+  Mul  I                  v         = v                           -- 1 * v       = v
+  Mul  u            (Inv  I)        = u                           -- u / 1       = u
+  Mul (u :*: Inv v)       v         = u                           -- u / v * v   = u
+  Mul (u :*: v)     (Inv  v)        = u                           -- u * v / v   = u
+  Mul (u :*: v)     (Inv  u)        = v                           -- u * v / u   = v
+  Mul  u                 (v :*: w)  = Mul (Mul u w) v             -- u * (v * w) = (u * w) * v
+  Mul  u            (Inv (v :*: w)) = Mul (Mul u (Inv w)) (Inv v) -- u / (v * w) = (u / w) / v
+  Mul (u :*: v)           w         = Mul u w :*: v               -- (u * v) * w = (u * w) * v
+  Mul  u                  v         = u :*: v                     -- u * v       = u * v
```

Comment:
I am incredibly pleased with how this turned out, especially in contrast to the code it replaced.